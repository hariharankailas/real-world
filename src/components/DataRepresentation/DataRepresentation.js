/***********************************************************************************************
 * File : DataRepresentation 
 * Desc : Fetch Comments and show the data in a tabular form 
 * Author: Hariharan Kailas
 * Date: 10-Jul-2021
 ***********************************************************************************************/
import React, { useState, Profiler } from 'react';
import { jsonplaceholder } from '../../api/jsonPlaceholder';
import './DataRepresentation.scss';
const DataRepresentation = ({ commentCb }) => {
    const [commentList, setCommentList] = useState([]);
    /***************************************
    * Desc: Load on comments from api
    * Date: 08-July-2021
    * ********************************/
    const loadComments = async () => {
        const response = await jsonplaceholder.get("/comments");
        setCommentList(response.data);
    }
    /***************************************
     * Desc: render table content
     * Date: 08-July-2021
     * ********************************/
    const renderTable = () => {
        return (
            <div className="table-responsive ">
                <table className="table table-striped table-hover">
                    <thead>
                        <tr>
                            <th scope="col">#</th>
                            <th scope="col">Name</th>
                            <th scope="col">Email</th>
                            <th scope="col">Description</th>
                        </tr>
                    </thead>
                    <tbody>
                        {renderTableRows()}
                    </tbody>
                </table>
                )
            </div>
        );
    }
    /***************************************
     * Desc: render table rows content
     * Date: 08-July-2021
     * ********************************/
    const renderTableRows = () => {
        return (
            <>
                {commentList.map(comment => <tr key={comment.id}>
                    <th scope="row">{comment.id}</th>
                    <td>{comment.name}</td>
                    <td>{comment.email}</td>
                    <td>{comment.body}</td>
                </tr>)
        
                }
            </>
        );
    }

    return (
        <div className="container">
            <div className="comment-control-container">
                <button type="button" id="fetchButton" className="btn btn-success btn-lg" onClick={ ()=> loadComments()}>
                    Fetch Comments
                </button>
                <button type="button"  id="clearButton" className="btn btn-danger btn-lg" onClick={ () => setCommentList([])}>
                    Clear
                    </button>
                </div>
            {commentList.length > 0 &&
                <Profiler id="commentsProfiler" onRender={commentCb}>
                    {renderTable()}
                </Profiler>
            }
            {commentList.length === 0 &&
                <h2 className="display-5 text-center">No Comments to Show</h2>
            }
        </div>
    );
}

export default DataRepresentation;