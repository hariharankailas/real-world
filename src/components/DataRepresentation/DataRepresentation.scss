.comment-control-container {
  display: flex;
  justify-content: space-around;
  margin: 2rem 0;
  button {
    width: 24rem;
  }
}

.table {
  width: 85%;
  margin: auto;
}

@media (max-width: 767.98px) {
  .comment-control-container {
    flex-wrap: wrap;
    button {
      width: 100%;
      margin: 1rem;
    }
  }
}
@media (min-width: 768px) and (max-width: 1024px) {
  .comment-control-container button {
    width: 20rem;
  }
}
