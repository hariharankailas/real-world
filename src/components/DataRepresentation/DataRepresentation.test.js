import React from 'react';
import { shallow } from 'enzyme';
import DataRepresentation from './DataRepresentation';

describe('LandingPage', () => {
  it('should render correctly', () => {
    const component = shallow(<DataRepresentation/>);
    expect(component).toMatchSnapshot();
  });

  it('on click of clear button set comment list to []', () => {
    const component = shallow(<DataRepresentation />);
    const clearCommentList = jest.spyOn(React, "useState");
    const setCommentList = jest.fn();
   clearCommentList.mockImplementation(commentList => [commentList, setCommentList]);
      component
    .find('#clearButton')
        .simulate('click');
           expect(setCommentList).toBeTruthy();

  });
  
  it('on click of fetch Comments set commentList State', () => {
    const component = shallow(<DataRepresentation />);
    const selectCommentList = jest.spyOn(React, "useState");
    const setCommentList = jest.fn();
    selectCommentList.mockImplementation(commentList => [commentList, setCommentList]);
    component.find('#fetchButton').simulate('click');
    expect(setCommentList).toBeTruthy();

  });
  
});