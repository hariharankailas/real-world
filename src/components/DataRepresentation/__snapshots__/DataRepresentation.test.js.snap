// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LandingPage should render correctly 1`] = `
<div
  className="container"
>
  <div
    className="comment-control-container"
  >
    <button
      className="btn btn-success btn-lg"
      id="fetchButton"
      onClick={[Function]}
      type="button"
    >
      Fetch Comments
    </button>
    <button
      className="btn btn-danger btn-lg"
      id="clearButton"
      onClick={[Function]}
      type="button"
    >
      Clear
    </button>
  </div>
  <h2
    className="display-5 text-center"
  >
    No Comments to Show
  </h2>
</div>
`;
